### PR TITLE
ci: Add build step to run-cluster

### PR DESCRIPTION
**Pull request by raunakab**
# Overview

The `run-cluster` workflow now also builds the commit by calling the `build-commit` workflow first.

### CODE REVIEW

**Comment by raunakab** [comment]
Example of a successful run:
https://github.com/Eventual-Inc/Daft/actions/runs/12403736836

File: .github/workflows/run-cluster.yaml [hunk 1]
```diff
@@ -45,12 +41,20 @@ on:
         default: ""
 
 jobs:
+  build-commit:
+    uses: ./.github/workflows/build-commit.yaml
+    with:
+      arch: x86
+      python_version: ${{ inputs.python_version }}
+    secrets:
+      ACTIONS_AWS_ROLE_ARN: ${{ secrets.ACTIONS_AWS_ROLE_ARN }}
```
Comment (jaychia) [inline .github/workflows/run-cluster.yaml:50]: Should this `job` not be optionally executed depending on whether:

1. User wants a specific published Daft version vs a ref
2. In the latter case, if a previously built wheel can be found

File: .github/workflows/run-cluster.yaml [hunk 2]
```diff
@@ -45,12 +41,20 @@ on:
...
+
   run-command:
     runs-on: [self-hosted, linux, x64, ci-dev]
-    timeout-minutes: 15 # Remove for ssh debugging
```
Comment (jaychia) [inline .github/workflows/run-cluster.yaml:50]: Seems you forgot to put this back in

File: .github/workflows/build-commit.yaml [hunk 3]
```diff
@@ -1,6 +1,24 @@
...
+        required: true
+        type: string
+      python_version:
+        description: The version of Python to use
+        required: true
+        type: string
+    secrets:
+      ACTIONS_AWS_ROLE_ARN:
+        description: The AWS Role ARN for secrets access
+        required: true
```
Comment (jaychia) [inline .github/workflows/build-commit.yaml:17]: Why is this required when it wasn't before?

Comment (raunakab) [reply .github/workflows/build-commit.yaml:17]: It's required when the workflow is invoked as a dependent workflow. Here is the PR that failed when `build-commit` didn't contain that:

https://github.com/Eventual-Inc/Daft/actions/runs/12403362351

File: .github/workflows/run-cluster.yaml (hunk 2 above)
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:50]: Actually, I removed it on purpose. Certain scale-factors and questions take over 15min. Here is a workflow that was proceeding successfully until the timeout was hit and everything else was cancelled:

https://github.com/Eventual-Inc/Daft/actions/runs/12401178483

File: .github/workflows/build-commit.yaml (hunk 3 above)
Comment (jaychia) [reply .github/workflows/build-commit.yaml:17]: I see, ok yeah we should check also that this doesn't break just running the build-commit workflow by itself from the UI or anything

File: .github/workflows/build-commit.yaml [hunk 4]
```diff
@@ -1,6 +1,24 @@
...
+    outputs:
+      wheel_url:
+        description: The name of the built wheel
```
Comment (jaychia) [inline .github/workflows/build-commit.yaml:20]: Is this the URL or name?

File: .github/workflows/build-commit.yaml [hunk 5]
```diff
@@ -1,6 +1,24 @@
 name: build-commit
 
 on:
+  workflow_call:
+    inputs:
+      arch:
+        description: The machine architecture to build for
+        required: true
+        type: string
```
Comment (jaychia) [inline .github/workflows/build-commit.yaml:9]: Probably `choice` instead like your other workflow?

File: .github/workflows/build-commit.yaml [hunk 6]
```diff
@@ -1,6 +1,24 @@
...
+      python_version:
+        description: The version of Python to use
+        required: true
+        type: string
```
Comment (jaychia) [inline .github/workflows/build-commit.yaml:13]: Add a default?

File: .github/workflows/run-cluster.yaml (hunk 1 above)
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:50]: Yes, that was my initial approach. However, I wanted to just get something basic up for this PR.

Possible implementation:
- we have a `check-if-built` job that runs first
  - if it does, it:
    - sets `wheel_url` to whatever it found
    - sets `should_build` to false
  - if not, then:
    - sets `should_build` to true
- the `build` job only execute if if `${{ needs.check-if-built.outputs.should_build == 'true' }}`

The problem that I was anticipating is how do I consume the wheel_url in the steps in `run-cluster`? Right now, it's `${{ needs.build-commit.outputs.wheel_url }}`, but if that step doesn't execute, then how does it find the wheel_url?

Possibly `${{ needs.build-commit.outputs.wheel_url || needs.check-if-built.outputs.wheel_url }}`? Will need to look into that some more.

File: .github/workflows/build-commit.yaml (hunk 5 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:9]: Actually, `choice`s aren't allowed in `workflow_call`. They're only allowed in `workflow_dispatch`. The types allowed in `workflow_call` are only `bool, string, number`.

File: .github/workflows/build-commit.yaml (hunk 4 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:20]: This will be the full URL of the wheel.

Comment (raunakab) [reply .github/workflows/build-commit.yaml:20]: I will update the description

File: .github/workflows/build-commit.yaml (hunk 3 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:17]: Ran one right now. Looks all good. You can observe the workflow file contains the `secrets` input as well.

https://github.com/Eventual-Inc/Daft/actions/runs/12404404099

File: .github/workflows/build-commit.yaml (hunk 6 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:13]: Updated.

File: .github/workflows/build-commit.yaml (hunk 4 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:20]: Updated.

File: .github/workflows/build-commit.yaml (hunk 5 above)
Comment (raunakab) [reply .github/workflows/build-commit.yaml:9]: Relevant docs:
https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype

File: .github/workflows/run-cluster.yaml [hunk 7]
```diff
@@ -3,10 +3,6 @@ name: run-cluster
 on:
   workflow_dispatch:
     inputs:
-      daft_wheel_url:
-        description: Daft python-wheel URL
-        type: string
-        required: false
       daft_version:
         description: Daft version (errors if both this and "Daft python-wheel URL" are provided)
```
Comment (jaychia) [inline .github/workflows/run-cluster.yaml:7]: (Proposal)

This can be changed to allow passing in one of:

1. Daft published version (maybe enforce that it starts with a `v`)
2. Branch/Git SHA (then we have to build)

**Comment by raunakab** [comment]
Example of a run that does *not* specify `daft-version`:
https://github.com/Eventual-Inc/Daft/actions/runs/12408535620

Example of a run that *does* specify `daft-version`:
https://github.com/Eventual-Inc/Daft/actions/runs/12408694177

File: .github/workflows/run-cluster.yaml (hunk 1 above)
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:51]: Used the `if: always()`. Works well.

File: .github/workflows/run-cluster.yaml (hunk 7 above)
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:7]: Scoped down to just either running with the:
1. current SHA (using `${{ github.sha }}`)
2. daft-version (by specifying using the input parameter)

File: .github/workflows/run-cluster.yaml [hunk 8]
```diff
@@ -3,10 +3,6 @@ name: run-cluster
 on:
   workflow_dispatch:
     inputs:
-      daft_wheel_url:
-        description: Daft python-wheel URL
-        type: string
-        required: false
```
[the remark this answers is not in the document]
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:9]: What we could do is leave in the `daft-wheel-url` parameter. The logic would look like:

```yaml
jobs:
  build-commit:
    if: ${{ inputs.daft_version == '' && inputs.daft_wheel_url == '' }}
    ...
    
  run-command:
    needs: build-commit
    if: always()
    steps:
      - name: Step xyz
        run: |
          python templatize_ray_config.py \
            --daft-wheel-url=${{ needs.build-commit.outputs.wheel_url || inputs.daft_wheel_url || '' }}
            --daft-version=${{ inputs.daft_version }}
            ...
```

Thus:
- if daft-version is specified, the `build-commit` job is skipped (and templatize is run with the daft-version is run with the daft_version input param)
- if daft-wheel-url is specified, the `build-commit` job is skipped again (and templatize is run with the daft_wheel_url input param)
- if neither are specified, the `build-commit` job is run (and templatize is run with the newly built wheel_url)
- if both are specified, the `build-commit` job is skipped (and templatize is run with daft-version and daft-wheel-url, and it throws an error)

[the remark this answers is not in the document]
Comment (jaychia) [reply .github/workflows/run-cluster.yaml:9]: Ok cool

File: .github/workflows/run-cluster.yaml [hunk 9]
```diff
@@ -45,12 +41,22 @@ on:
...
+    with:
+      arch: x86
+      python_version: ${{ inputs.python_version }}
+    secrets:
+      ACTIONS_AWS_ROLE_ARN: ${{ secrets.ACTIONS_AWS_ROLE_ARN }}
+
   run-command:
     runs-on: [self-hosted, linux, x64, ci-dev]
-    timeout-minutes: 15 # Remove for ssh debugging
+    if: always()
```
[the remark this answers is not in the document]
Comment (raunakab) [reply .github/workflows/run-cluster.yaml:62]: Updated.

File: .github/workflows/run-cluster.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (raunakab) [reply .github/workflows/run-cluster.yaml]: Added back the `daft_wheel_url`. Running 3 jobs now to test the various scenarios.

[the remark this answers is not in the document]
Comment (raunakab) [reply .github/workflows/run-cluster.yaml]: Support for daft-wheel-url still works.
